### PR TITLE
ci: Add integration tests for secret & configmap propagation

### LINKED ISSUE #8015: ci: test: Add integration tests for secret & configmap propagation
In https://github.com/kata-containers/kata-containers/pull/7211, we added support for propagation of secrets/configmaps/... in an environment without shared fs.

The idea behind this is that if we make a change to a kubernetes secret on the host then the watcher will update it in the peer pod as well. We should create an integration tests for this to check if this works.

There is also a complication with this around sealed secrets feature and if this impacts this behaviour, so we want to ensure it runs if `sealed_secret_enabled is true, or false.

### Acceptance Criteria

_**Scenario: secrets update**_
**Given** We have a Kubernetes deployment with a pod running on in an environment with sharedfs=none and a kubernetes secret 
**When** I edit the secret to a new value
**Then** I when I exec into the pod to check the value, I can see the secret change has been propogated into the pod

_**Scenario: config map update**_
**Given** We have a Kubernetes deployment with a pod running on in an environment with sharedfs=noneand a kubernetes configmap
**When** I edit the configmap to a new value
**Then** I when I exec into the pod to check the value, I can see the config map change has been propogated into the pod

### PR DESCRIPTION
**Pull request by ajaypvictor**
Enhanced `k8s-configmap.bats` and `k8s-credentials-secrets.bats` to verify that ConfigMap and Secret updates propagate to volume-mounted pods.

Changes:
- `k8s-configmap.bats`: Added volume mount test and propagation verification
- `k8s-credentials-secrets.bats`: Added propagation verification
- Improved cleanup by moving resource deletion to `teardown()`

Tests use retry logic to validate propagation within kubelet sync period and maximize reuse of existing YAML resources.

Fixes #8015

### CODE REVIEW

File: tests/integration/kubernetes/k8s-configmap-propagation.bats [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+load "${BATS_TEST_DIRNAME}/lib.sh"
+load "${BATS_TEST_DIRNAME}/tests_common.sh"
+
+setup() {
+	setup_common
+	get_pod_config_dir
+
+	if ! is_confidential_runtime_class; then
+		skip "Test not supported for ${KATA_HYPERVISOR}."
+	fi
```
Comment (stevenhorsman) [inline tests/integration/kubernetes/k8s-configmap-propagation.bats]: I think you need this block right at the start of `setup` and `teardown` so we don't run either of them.

**Comment by BbolroC** [comment]
I am running the test for this PR on IBM SEL machine.

File: tests/integration/kubernetes/k8s-configmap.bats [hunk 2]
```diff
@@ -43,15 +46,75 @@ setup() {
...
+	# Cleanup
 	kubectl delete pod "$pod_name"
 	kubectl delete configmap "$config_name"
+	delete_tmp_policy_settings_dir "${policy_settings_dir}"
+}
+
+@test "ConfigMap propagation in pod with sharedfs=none" {
+	if ! is_confidential_hardware; then
+		skip "Test requires shared_fs=none (confidential hardware: ${KATA_HYPERVISOR})"
+	fi
```
Comment (sprt) [inline tests/integration/kubernetes/k8s-configmap.bats]: Can we simply merge this test case with the original one and remove this conditional? This way we also test propagation when `shared_fs != none`. 🙂

Same for secrets.

**Comment by BbolroC** [comment]
The latest push passed on IBM SEL. Thanks!

**Comment by stevenhorsman** [comment]
@ajaypvictor - can you rebase this PR please? Thanks!

File: tests/integration/kubernetes/k8s-configmap.bats [hunk 3]
```diff
@@ -11,47 +11,87 @@ load "${BATS_TEST_DIRNAME}/tests_common.sh"
...
+	# If we get here, propagation failed
+	info "ConfigMap propagation test failed"
+	return 1
+}
+
+teardown() {
+	# Cleanup resources
+	kubectl delete pod config-env-test-pod --ignore-not-found=true
+	kubectl delete pod configmap-volume-test-pod --ignore-not-found=true
+	kubectl delete configmap test-configmap --ignore-not-found=true
```
[the remark this answers is not in the document]
Comment (ajaypvictor) [reply tests/integration/kubernetes/k8s-configmap.bats]: These variables are defined on the test stage which is under `@test`, might need to export the variables as,
```
export config_name="test-configmap"
export pod_env_name="config-env-test-pod"
export pod_volume_name="configmap-volume-test-pod"
```
for using on `teardown()` stage?

EDIT: will move this logic to setup stage, so that teardown could access it.

**Comment by stevenhorsman** [comment]
@ajaypvictor - the new editorconfig is failing due to a lot of trailing whitespace in these files. Are you able to fix those up please.

**Comment by ajaypvictor** [comment]
> @ajaypvictor - the new editorconfig is failing due to a lot of trailing whitespace in these files. Are you able to fix those up please.

Thanks for pointing this Steve, fixed it..

File: tests/integration/kubernetes/k8s-configmap.bats [hunk 4]
```diff
@@ -10,48 +10,84 @@ load "${BATS_TEST_DIRNAME}/../../common.bash"
...
+
+	# Test 2: Create a pod that consumes the ConfigMap via volume mount
+	kubectl create -f "${pod_volume_yaml_file}"
+	kubectl wait --for=condition=Ready --timeout=$timeout pod "$pod_volume_name"
+
+	# Verify initial value from volume
+	grep_pod_exec_output "${pod_volume_name}" "value-1" "${check_config_exec_command[@]}"
+
+	# Update ConfigMap to test propagation
+	kubectl patch configmap "${config_name}" -p '{"data":{"data-1":"updated-value-1"}}'
```
Comment (stevenhorsman) [inline tests/integration/kubernetes/k8s-configmap.bats]: nit: We could extract `updated-value-1` as a separate variable e.g. updatedValue

File: tests/integration/kubernetes/k8s-configmap.bats [hunk 5]
```diff
@@ -10,48 +10,84 @@ load "${BATS_TEST_DIRNAME}/../../common.bash"
...
 	kubectl create -f "${configmap_yaml_file}"
 
 	# View the values of the keys
 	kubectl get configmaps $config_name -o yaml | grep -q "data-"
 
-	# Create a pod that consumes the ConfigMap
-	kubectl create -f "${pod_yaml_file}"
+	# Test 1: Create a pod that consumes the ConfigMap via environment variables
+	kubectl create -f "${pod_env_yaml_file}"
+	kubectl wait --for=condition=Ready --timeout=$timeout pod "$pod_env_name"
```
Comment (stevenhorsman) [inline tests/integration/kubernetes/k8s-configmap.bats]: It would be good to get variable braces everywhere here as that's our repo's prefered style e.g.
```suggestion
	kubectl wait --for=condition=Ready --timeout=${timeout pod} "${pod_env_name}"
```

File: tests/integration/kubernetes/k8s-configmap.bats [hunk 6]
```diff
@@ -10,48 +10,84 @@ load "${BATS_TEST_DIRNAME}/../../common.bash"
...
+
+	# Check env variables
+	grep_pod_exec_output "${pod_env_name}" "KUBE_CONFIG_1=value-1" "${exec_command[@]}"
```
Comment (stevenhorsman) [inline tests/integration/kubernetes/k8s-configmap.bats:53]: nit: Extracting `value-1` as original value would be nice too?

File: tests/integration/kubernetes/k8s-configmap.bats (hunk 4 above)
Comment (ajaypvictor) [reply tests/integration/kubernetes/k8s-configmap.bats]: updated, thanks for pointing this out..

File: tests/integration/kubernetes/k8s-configmap.bats (hunk 5 above)
Comment (ajaypvictor) [reply tests/integration/kubernetes/k8s-configmap.bats]: updated, thanks for pointing this out..

File: tests/integration/kubernetes/k8s-configmap.bats (hunk 6 above)
Comment (ajaypvictor) [reply tests/integration/kubernetes/k8s-configmap.bats:53]: updated, thanks for pointing this out..

File: tests/integration/kubernetes/k8s-configmap.bats [hunk 7]
```diff
@@ -10,48 +10,86 @@ load "${BATS_TEST_DIRNAME}/../../common.bash"
...
+
+	# Test 1: Create a pod that consumes the ConfigMap via environment variables
+	kubectl create -f "${pod_env_yaml_file}"
+	kubectl wait --for=condition=Ready --timeout="${timeout}" pod "${pod_env_name}"
+
+	# Check env variables
+	grep_pod_exec_output "${pod_env_name}" "KUBE_CONFIG_1=${original_value}" "${exec_command[@]}"
+	grep_pod_exec_output "${pod_env_name}" "KUBE_CONFIG_2=value-2" "${exec_command[@]}"
+
+	# Test 2: Create a pod that consumes the ConfigMap via volume mount
```
Comment (sprt) [inline tests/integration/kubernetes/k8s-configmap.bats]: My apologies, I had missed that test 1 uses env variables and test 2 uses volumes. It makes more sense to split this into separate `@test` cases then.

File: tests/integration/kubernetes/k8s-credentials-secrets.bats [hunk 8]
```diff
@@ -37,45 +47,65 @@ setup() {
...
 	pod_exec "${pod_name}" "${pod_exec_command[@]}" | grep -w "username"
 
-	# Create a pod that has access to the secret data through environment variables
-	kubectl create -f "${pod_env_yaml_file}"
+	# Verify initial secret value
+	grep_pod_exec_output "${pod_name}" "${original_username}" "${check_secret_exec_command[@]}"
 
-	# Check pod creation
-	kubectl wait --for=condition=Ready --timeout=$timeout pod "$second_pod_name"
+	# Test 2: Create a pod that has access to the secret data through environment variables
```
Comment (sprt) [inline tests/integration/kubernetes/k8s-credentials-secrets.bats]: Same comment as for configMaps.

File: tests/integration/kubernetes/k8s-configmap.bats (hunk 7 above)
Comment (ajaypvictor) [reply tests/integration/kubernetes/k8s-configmap.bats]: updated, thanks for the suggestion..

File: tests/integration/kubernetes/k8s-credentials-secrets.bats (hunk 8 above)
Comment (ajaypvictor) [reply tests/integration/kubernetes/k8s-credentials-secrets.bats]: updated, thanks for the suggestion..
